vcd/resource_vcd_vm_internal_disk.go

```diff
@@ -0,0 +1,477 @@
+				Description: "The size of the disk in MB.",
+			},
+			"bus_number": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The number of the SCSI or IDE controller itself.",
+			},
+			"unit_number": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The device number on the SCSI or IDE controller of the disk.",
+			},
+			"thin_provisioned": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Specifies whether the disk storage is pre-allocated or allocated on demand.",
+			},
+			"iops": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+				Description: "Specifies the IOPS for the disk.",
+			},
+			"storage_profile": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Storage profile to override the VM default one",
+			},
+		},
+	}
+}
+
+var internalDiskBusTypes = map[string]string{
+	"ide":         "1",
+	"parallel":    "3",
+	"sas":         "4",
+	"paravirtual": "5",
+	"sata":        "6",
+}
+var internalDiskBusTypesFromValues = map[string]string{
+	"1": "ide",
+	"3": "parallel",
+	"4": "sas",
+	"5": "paravirtual",
+	"6": "sata",
+}
+
+var vmStatusBefore string
+
+// resourceVmInternalDiskCreate creates an internal disk for VM
+func resourceVmInternalDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	vcdClient.lockParentVm(d)
+	defer vcdClient.unLockParentVm(d)
+
+	vm, vdc, err := getVm(vcdClient, d)
+	if err != nil {
+		return err
+	}
+
+	var storageProfilePrt *types.Reference
+	var overrideVmDefault bool
+
+	if storageProfileName, ok := d.GetOk("storage_profile"); ok {
+		storageProfile, err := vdc.FindStorageProfileReference(storageProfileName.(string))
+		if err != nil {
+			return fmt.Errorf("[vm creation] error retrieving storage profile %s : %s", storageProfileName, err)
+		}
+		storageProfilePrt = &storageProfile
+		overrideVmDefault = true
+	} else {
+		storageProfilePrt = vm.VM.StorageProfile
+		overrideVmDefault = false
+	}
+
+	iops := int64(d.Get("iops").(int))
+	// value is required but not treated.
+	isThinProvisioned := true
+
+	diskSetting := &types.DiskSettings{
+		SizeMb:              int64(d.Get("size_in_mb").(int)),
+		UnitNumber:          d.Get("unit_number").(int),
+		BusNumber:           d.Get("bus_number").(int),
+		AdapterType:         internalDiskBusTypes[d.Get("bus_type").(string)],
+		ThinProvisioned:     &isThinProvisioned,
+		StorageProfile:      storageProfilePrt,
+		Iops:                &iops,
+		VirtualQuantityUnit: "byte",
+		OverrideVmDefault:   overrideVmDefault,
+	}
+
+	err = powerOffIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	diskId, err := vm.AddInternalDisk(diskSetting)
+	if err != nil {
+		return fmt.Errorf("error updating VM disks: %s", err)
+	}
+
+	d.SetId(diskId)
+
+	err = powerOnIfNeeded(d, vm)
```

Comment:
Shouldn't this actually check what was the state before line 160 (where `powerOffIfNeeded` is called).
_Scenario_: the VM is already powered off and I want to keep it that way after adding a disk (especially if I have `power_on=false` in `vcd_vapp_vm` which we can't access here).

Basically it is the same comment as in line 245.

Reply:
It's handled inside that function

Reply:
Solved now. For reference - previously it shared state in global variable so that was risky if many disks are being modified. Now it uses local state tracking.